stint/io.nim

```diff
@@ -512,3 +512,11 @@ func toByteArrayBE*[bits: static[int]](n: StUint[bits]): array[bits div 8, byte]
       let n_ptr {.restrict.} = cast[ptr array[N, byte]](n.unsafeAddr)
       for i in 0 ..< N:
         result[N-1 - i] = n_ptr[i]
+
+func hash*(num: UintImpl|IntImpl): Hash =
+  mixin hash
+  hash(num.hi) xor hash(num.lo)
```

Comment:
ostensibly, there are better ways to combine hashes: https://stackoverflow.com/questions/4948780/magic-number-in-boosthash-combine

also, this is a periodic reminder that calling functions recursively on stint generates extremely inefficient code (n^2 code size on n depth) - we need to fix this across the board though, so it's not a blocker for this PR

Reply:
Anyone should feel free to improve on this. I needed it in order to make progress on the discv5 integration where Stint numbers are used as IDs for nodes in the network.

I briefly considered using something different than xor, but it seemed to me that there is value is preserving one of the properties of normal `int` values - if you use the first N natural numbers in a hash table, you'll get very efficient non-colliding indexing.

Reply:
I somehow suspect the boost authors spent more time than "briely" on this problem - it "seems" reasonable to use their work as a baseline unless you have a researched solution which is demonstrably better, no?

Reply:
reason to care being that unexpected linear performance in hash tables is a real pita..
also, tests?

Reply:
Don't get me wrong. I don't oppose using the more sophisticated method. I just brought to the table the additional perspective that the stint values are integer-like - even using their least significant word as the hash value would be equivalent to doing `X mod INT_MAX` which may be a reasonable solution for a hash table (considering that regular integers get roughly the same treatment).

Reply:
FWIW, I've seen the randomizing trick to be used with regular integers as well.

Reply:
`xor` alone seems not random enough, at least some shifts in the style of xorshift and xoshiro and xoroshiro would probably much more robust.

http://prng.di.unimi.it/

See xoshiro512: http://prng.di.unimi.it/xoshiro512starstar.c

```C
/*  Written in 2018 by David Blackman and Sebastiano Vigna (vigna@acm.org)

To the extent possible under law, the author has dedicated all copyright
and related and neighboring rights to this software to the public domain
worldwide. This software is distributed without any warranty.

See <http://creativecommons.org/publicdomain/zero/1.0/>. */

#include <stdint.h>
#include <string.h>

/* This is xoshiro512** 1.0, one of our all-purpose, rock-solid generators
   with increased state size. It has excellent (about 1ns) speed, a state
   (512 bits) that is large enough for any parallel application, and it
   passes all tests we are aware of.

   For generating just floating-point numbers, xoshiro512+ is even faster.

   The state must be seeded so that it is not everywhere zero. If you have
   a 64-bit seed, we suggest to seed a splitmix64 generator and use its
   output to fill s. */

static inline uint64_t rotl(const uint64_t x, int k) {
	return (x << k) | (x >> (64 - k));
}


static uint64_t s[8];

uint64_t next(void) {
	const uint64_t result = rotl(s[1] * 5, 7) * 9;

	const uint64_t t = s[1] << 11;

	s[2] ^= s[0];
	s[5] ^= s[1];
	s[1] ^= s[2];
	s[7] ^= s[3];
	s[3] ^= s[4];
	s[4] ^= s[5];
	s[0] ^= s[6];
	s[6] ^= s[7];

	s[6] ^= t;

	s[7] = rotl(s[7], 21);

	return result;
}
```

I agree that linear perf in hashtables is crippling: https://github.com/nim-lang/Nim/issues/11581

Reply:
indeed - when making a general purpose integer hash function, you distinctly _do not_ wish to do modulo..